recipes-kernel/linux/5.15/patches/blktap2.patch

```diff
@@ -1039,14 +1044,6 @@ PATCHES
 +
 +	bdev = bdget_disk(gd, 0);
 +
-+	err = !mutex_trylock(&bdev->bd_mutex);
-+	if (err) {
-+		/* NB. avoid a deadlock. the last opener syncs the
-+		 * bdev holding bd_mutex. */
-+		err = -EBUSY;
-+		goto out_nolock;
-+	}
```

Comment:
Sorry, I didn't notice this earlier - what is the rationale that it is safe to remove this locking?

Looks like linux commit a8698707a1835be3abd12a3b28079a80999f8dee
"block: move bd_mutex to struct gendisk" moved it up a level.  Looking at that commit, del_gendisk tries to take the lock, and we call that below, so maybe that is part of why is needs to be dropped?

I think the point of the lock is to serialize a race between the block device cleaning up and a second process opening it.  I also thought it should be held to look at the bdev data like bd_openers.

Maybe not for now, but commit 05d69d950d9d84218fc9beafd02dea1f6a70e09e
"xen-blkfront: sanitize the removal state machine" cleaned up blkfront for its removal case.  Following that may be good idea for blktap2 as well.

Reply:
I think you're right with the bd_mutex, though perhaps the original code here was piggybacking on it and it was really needing a completely separate lock.  We don't want to double-lock, so replacing this call with something like ```bdev->bd_disk->open_mutex``` isn't going to work.  I have to imagine this is a pointer to the same lock that's in the ```gendisk``` object ```gd->open_mutex``` that ```del_gendisk``` grabs, like you mentioned.

If we still want to serialize access to the ```blktap_device_destroy``` function, which is effectively what ```bd_mutex``` was being used for originally, I agree, we need to replace it with something else.  Let me see what I can do

Reply:
Actually, can we just move your ```mutex_lock(&tapdev->lock)``` up?

Reply:
I'd rather not touch 05d69d950d9d84218fc9beafd02dea1f6a70e09e at the moment.  I will try the trylock for open_mutex and see what happens.

Reply:
That sounds reasonable.

Reply:
using the trylock on open_mutex led to deadlock.  moving the tapdev->lock further up seems to work much better.

Reply:
Locally, I tried `mutex_trylock(&bdev->bd_disk->open_mutex);` and can confirm it deadlocks :(

Reply:
:( indeed